sdk/schemaregistry/Azure.Data.SchemaRegistry/src/Avro/SchemaRegistryAvroObjectSerializer.cs

```diff
@@ -0,0 +1,241 @@
+        {
+            if (typeof(ISpecificRecord).IsAssignableFrom(type))
+            {
+                return SupportedType.SpecificRecord;
+            }
+
+            if (typeof(GenericRecord).IsAssignableFrom(type))
+            {
+                return SupportedType.GenericRecord;
+            }
+
+            throw new ArgumentException($"Type {type.Name} is not supported for serialization operations.");
+        }
+
+        private string GetSchemaId(Schema schema, CancellationToken cancellationToken)
+        {
+            var schemaProperties = _options.AutoRegisterSchemas
+                ? _client.RegisterSchema(_groupName, schema.Fullname, SerializationType.Avro, schema.ToString(), cancellationToken)
+                : _client.GetSchemaId(_groupName, schema.Fullname, SerializationType.Avro, schema.ToString(), cancellationToken);
+            return schemaProperties.Value.Id;
+        }
+
+        private async Task<string> GetSchemaIdAsync(Schema schema, CancellationToken cancellationToken)
+        {
+            var schemaProperties = await (_options.AutoRegisterSchemas
+                    ? _client.RegisterSchemaAsync(_groupName, schema.Fullname, SerializationType.Avro, schema.ToString(), cancellationToken)
+                    : _client.GetSchemaIdAsync(_groupName, schema.Fullname, SerializationType.Avro, schema.ToString(), cancellationToken))
+                .ConfigureAwait(false);
+            return schemaProperties.Value.Id;
+        }
+
+        private static DatumWriter<object> GetWriterAndSchema(object value, SupportedType supportedType, out Schema schema)
+        {
+            switch (supportedType)
+            {
+                case SupportedType.SpecificRecord:
+                    schema = ((ISpecificRecord)value).Schema;
+                    return new SpecificDatumWriter<object>(schema);
+                case SupportedType.GenericRecord:
+                    schema = ((GenericRecord)value).Schema;
+                    return new GenericDatumWriter<object>(schema);
+                default:
+                    throw new ArgumentException($"Invalid supported type value: {supportedType}");
+            }
+        }
+
+        /// <inheritdoc />
+        public override void Serialize(Stream stream, object value, Type inputType, CancellationToken cancellationToken)
+        {
+            Argument.AssertNotNull(stream, nameof(stream));
+            Argument.AssertNotNull(value, nameof(value));
+            Argument.AssertNotNull(inputType, nameof(inputType));
+
+            var supportedType = GetSupportedTypeOrThrow(inputType);
+            var writer = GetWriterAndSchema(value, supportedType, out var schema);
+            var schemaId = GetSchemaId(schema, cancellationToken);
+
+            var binaryEncoder = new BinaryEncoder(stream);
+            stream.Write(s_emptyRecordFormatIndicator, 0, 4);
+            stream.Write(Encoding.UTF8.GetBytes(schemaId), 0, 32);
```

Comment:
Do we intend to include the BOM here? I think Encoding.Utf8 includes the BOM by default.
In [BinaryData](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs#L39), we do not include the BOM.

Also, should we make this a static for perf?

Reply:
Our doc for the format did not mention preamble (BOM). `GetBytes()` does not include the BOM. You have to use `GetPreamble()` and concat with `GetBytes()`. I have, however, added a static readonly field for the encoding. But, I only did that because I don't know if the `Encoding.UTF8` property is cached or not (meaning, it isn't creating a new instance every Get).

Reply:
I've gotten confirmation that BOM is not part of the message format.

Reply:
The issue is that Encoding.Utf8 will include the BOM. So if you want to exclude it, I think you will need to specify that in the ctor as we do in BinaryData.

Reply:
Nevermind, if you use the ctor it is excluded by default.